src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs

```diff
@@ -63,6 +82,38 @@ public async Task<string> GetSecretValue(Uri secretUri, CancellationToken cancel
             return secretValue;
         }
 
+        internal bool AnyExpiredSecrets()
+        {
+            bool shouldRefreshKeyVaultSecrets = false;
+            List<CachedKeyVaultSecret> secretsToBeRemovedFromCache = new List<CachedKeyVaultSecret>();
+
+            foreach (var cachedSecret in _cachedKeyVaultSecrets)
+            {
+                // Skip the refresh for this key vault secret if it has no expiration time or if it hasn't expired yet
+                if (cachedSecret.ExpiresOn == null || DateTimeOffset.UtcNow < cachedSecret.ExpiresOn)
+                {
+                    continue;
+                }
+
+                // Remove the cached Key Vault secret for this key
+                secretsToBeRemovedFromCache.Add(new CachedKeyVaultSecret(cachedSecret.Key));
+                shouldRefreshKeyVaultSecrets = true;
+            }
+
+            secretsToBeRemovedFromCache.ForEach(secret => RemoveExpiredSecretFromCache(secret.Key));
```

Comment:
>RemoveExpiredSecretFromCache [](start = 58, length = 28)

It's a bit hard to follow a query method that also modifies state. I'd suggest separating it somehow. I don't think we really need to remove expired secrets from the cache until we query for a secret again.

Reply:
Moved the deletion logic out of this query method as discussed offline. Now `GetSecretValue` will take care of updating/removing stale secrets.